src/datasets/vision/imagenet_reader/preprocess.jl

```diff
@@ -0,0 +1,34 @@
+# Image preprocessing for ImageNet models.
+# Code adapted from Metalhead 0.5.3's utils.jl
+
+# Coefficients taken from PyTorch's ImageNet normalization code
+const PYTORCH_MEAN = [0.485f0, 0.456f0, 0.406f0]
+const PYTORCH_STD = [0.229f0, 0.224f0, 0.225f0]
+
+normalize_pytorch(x) = (x .- PYTORCH_MEAN) ./ PYTORCH_STD
+inv_normalize_pytorch(x) = x .* PYTORCH_STD .+ PYTORCH_MEAN
```

Comment:
I would drop the `pytorch` prefix/suffix and use something else. The comments can stay. If PyTorch isn't the only library that does this preprocessing, then it makes sense to represent that with more general names. If different libraries are providing different preprocessing functionality for ImageNet (or not providing any), then I'd argue there is no canonical default set of ImageNet transformations and this code (aside from maybe the descriptive stats) shouldn't be in MLDatasets.

Reply:
Good point. Since this is just an internal function used by `default_preprocess`, I would suggest either `_normalize` or `default_normalize`. The appeal of using these coefficients as defaults is that they should work out of the box with pre-trained vision models from [Metalhead.jl](https://github.com/FluxML/Metalhead.jl/).

Reply:
Wait, so do other libraries provide this functionality in their ImageNet dataset APIs? I checked https://www.tensorflow.org/datasets/catalog/imagenet2012 and it has no mention of preprocessing, so is PyTorch the only library that does this? If so, I would vote to remove the preprocessing functions as mentioned above.

Reply:
If I am not wrong, these normalization values depend on the model you are using. Also,  none of the existing vision datasets have preprocessing functions. These functions are ideally handled by data preprocessing libraries/modules.

Reply:
The norm values should not be model-specific. They're derived directly from the data before any model is involved.

Reply:
For reference, pytorch includes normalization constants as part of the "model weights" ([see here](https://pytorch.org/vision/stable/models/generated/torchvision.models.resnet34.html#torchvision.models.ResNet34_Weights)):
> The inference transforms are available at ResNet34_Weights.IMAGENET1K_V1.transforms and perform the following preprocessing operations: Accepts PIL.Image, batched (B, C, H, W) and single (C, H, W) image torch.Tensor objects. The images are resized to resize_size=[256] using interpolation=InterpolationMode.BILINEAR, followed by a central crop of crop_size=[224]. Finally the values are first rescaled to [0.0, 1.0] and then normalized using mean=[0.485, 0.456, 0.406] and std=[0.229, 0.224, 0.225].

Another famous pretrained-models library, the [timm library](https://github.com/rwightman/pytorch-image-models/blob/main/timm/data/constants.py), includes the normalization weights as `from timm.data.constraints import IMAGENET_DEFAULT_MEAN, IMAGENET_DEFAULT_STD`.

Reply:
In the pytorch case, notice however that although the transformations are stored in the "model weights", the mean and std is the same across models ([see e.g. the mobilenet model](https://pytorch.org/vision/stable/models/generated/torchvision.models.mobilenet_v3_large.html#torchvision.models.MobileNet_V3_Large_Weights)).

In a similar spirit, I would definitely defend the decision of shipping the set of transformations (cropping, interpolation, linear transformation, etc) as part of the dataset. However I agree with the very first point that the name `transformation_pytorch` isn't really precise, although I think it is fair to link to the corresponding transformations for tensorflow, pytorch, and/or the timm library in a related comment.

Reply:
PyTorch also lumps code for pretrained models, data augmentations and datasets into one library, I don't think we need to follow their every example :)

> In a similar spirit, I would definitely defend the decision of shipping the set of transformations (cropping, interpolation, linear transformation, etc) as part of the dataset.

This is precisely why I asked about what other libraries are doing. If nobody else is shipping the same set of transformations, then they can hardly be considered canonical for ImageNet. That doesn't mean we should never ship helpers to create common augmentation pipelines, but that is better served by packages which have access to efficient augmentation libraries (e.g. Augmentor, DataAugmentation) and not by some unoptimized implementation which is simultaneously more general (because it's applicable to other datasets) and less general (because many papers using ImageNet do not use these augmentations) than the dataset it's been attached to.

Reply:
Let's just apply the `channelview` and permute transformation by default here,
and make the (permuted) mean and std values be part of the type

Reply:
I have also taken a look at Keras' ImageNet utilities. While these normalization constants are used in many places throughout [torchvision](https://github.com/search?q=repo%3Apytorch%2Fvision%200.406&type=code) and [PyTorch](https://github.com/search?q=repo%3Apytorch%2Fpytorch+0.406&type=code), it looks like TensorFlow and Keras do indeed [use their own constants](https://github.com/keras-team/keras/blob/db11e9674157c9125c9215b3bd4928675cf5af53/keras/applications/imagenet_utils.py#L200-L219).

I agree with @ToucheSir's sentiment
> If nobody else is shipping the same set of transformations, then they can hardly be considered canonical for ImageNet.

However, this point can be drawn even further, as nothing about ImageNet is truly canonical.
To give some examples (some of which have previously been discussed):

1) There is no canonical reason why Images have to be loaded in 224 x 224 format.
2) There is no canonical reason to apply the resizing algorithm JpegTurbo.jl uses when calling `jpeg_decode` with a `preferred_size`.
3) There is no canonical way of sorting class labels.  Some sort by WordNet ID (e.g. PyTorch), others don't.

### Getting this merged
So make this dataloader as "unopinionated" as possible, we could just make it a very thin wrapper around `FileDataset` which just loads metadata. This would require the user to pass a `loadfn` which handles the transformation from file path to array. Class ordering could be handled using a `sort_by_wnid=True`  keyword argument and all new dependencies introduced in this PR could  be removed (ImageCore, JpegTurbo and StackViews).

### Future work
However, I do strongly feel like some package in the wider Julia ML / Deep Learning ecosystem should export `loadfn`s that are usable with Metalhead's PyTorch models out of the box. @lorenzoh previously proposed adding such functionality to DataAugmentation.jl in https://github.com/FluxML/Metalhead.jl/issues/117. 
Once this functionality is available somewhere, `ImageNet`'s docstring in MLDataset should be updated to showcase this common use-case.

Until this functionality exists, I would suggest adding a `"Home" => "Tutorials" => "ImageNet"` page to the MLDatasets docs which implements the current load function.

Reply:
> it looks like TensorFlow and Keras do indeed [use their own constants](https://github.com/keras-team/keras/blob/db11e9674157c9125c9215b3bd4928675cf5af53/keras/applications/imagenet_utils.py#L200-L219).

Nice find. I was not expecting that `mode == "torch"` conditional.

> However, this point can be drawn even further, as nothing about ImageNet is truly canonical. To give some examples (some of which have previously been discussed):
> 
>     1. There is no canonical reason why Images have to be loaded in 224 x 224 format.
> 
>     2. There is no canonical reason to apply the resizing algorithm JpegTurbo.jl uses when calling `jpeg_decode` with a `preferred_size`.
> 
>     3. There is no canonical way of sorting class labels.  Some sort by WordNet ID (e.g. PyTorch), others don't.

The difference here is that all three of those points can have a decent fallback without depending on external packages. Another argument is that more people will rely on these defaults than won't. I'm not sure augmentations pass that threshold.
I'm not saying users shouldn't be able to pass in a transformation function, but `identity` or some such seems a more defensible default. Indeed, the torchvision `ImageNet` class does not do any additional transforms by default, so we'd be deviating from every other library if we stuck with this default centre crop.